soc/arm/st_stm32/stm32l1/power.c

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (c) 2020 Abel Sensors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <power/power.h>
+#include <soc.h>
+#include <init.h>
+
+#include <stm32_ll_cortex.h>
+#include <stm32_ll_pwr.h>
+#include <stm32_ll_system.h>
+#include <stm32_ll_bus.h>
+
+#include <logging/log.h>
+
+LOG_MODULE_DECLARE(soc, CONFIG_SOC_LOG_LEVEL);
+
+/**
+ * @brief Put processor into a power state.
+ *
+ * This function implements the SoC specific details necessary
+ * to put the processor into available power states.
+ */
+void pm_power_state_set(struct pm_state_info info)
+{
+	if (info.state != PM_STATE_SUSPEND_TO_IDLE) {
+		LOG_DBG("Unsupported power state %u", info.state);
+		return;
+	}
+
+	switch (info.substate_id) {
+	case 0:
+		/* this corresponds to the STOP mode: */
+
+#ifdef CONFIG_DEBUG
+		/* Enable the Debug Module during STOP mode */
+		LL_DBGMCU_EnableDBGStopMode();
+#endif /* CONFIG_DEBUG */
+
+		/** Request to enter STOP mode
+		 * Following procedure described in STM32L1xx Reference Manual
+		 * See PWR part, section Low-power modes, STOP mode
+		 */
+
+		/* Enable ultra low power mode */
+		LL_PWR_EnableUltraLowPower();
+
+		/** Set the regulator to low power before setting MODE_STOP.
+		 * If the regulator remains in "main mode",
+		 * it consumes more power without providing any additional feature.
+		 */
+		LL_PWR_SetRegulModeLP(LL_PWR_REGU_LPMODES_LOW_POWER);
```

Comment:
I can see you don't call `LL_PWR_SetRegulModeLP(LL_PWR_REGU_LPMODES_MAIN);`
Is that done on purpose ?

Reply:
Yes, since this decreases the consumption further. Is there anything specific against having this in the STM32L1x-wide sleep mode selection?

Reply:
My question wasn't clear enough:
I can see you don't call `LL_PWR_SetRegulModeLP(LL_PWR_REGU_LPMODES_MAIN);` **when exiting sleep mode**.
Is that done on purpose ?`

Reply:
When I check the reference manual [[here](https://www.st.com/resource/en/reference_manual/cd00240193-stm32l100xx-stm32l151xx-stm32l152xx-and-stm32l162xx-advanced-arm-based-32-bit-mcus-stmicroelectronics.pdf)] pg. 122, I see that the voltage regulator is only put in low-power mode during deepsleep/sleep/low-power run mode. If I understand correctly it shouldn't impact the voltage regulator during regular run (so after exiting sleep mode), since it also states "_The regulator goes back to main mode when the CPU exits from these modes_". Was that indeed your concern?

Reply:
Ok. Indeed that's a good thing to set this.
Though, I'd rather move this in `stm32_power_init` as current function is a transition function. We expect to see here things that are need to be set when entering sleep mode and because they are unset when exiting sleep mode.

Reply:
Okay, will do. Do want it to be in a switch case similar to the one used in `pm_power_state_set()`, since that clarifies that this is required specifically for the stop mode (in case something different is required for other low-power modes)?

Reply:
My point was to do it unconditionally and so in `stm32_power_init`. Since you do it when entering Low power mode and don't "undo it", then it can be done unconditionally at startup, no need to wait to enter LP mode.

Reply:
Yes I understand, but I want to ensure that these configurations are applicable for all non-run modes, in case more are added. `LL_PWR_EnableUltraLowPower()` for example isn't.

Reply:
The functions that are not applicable in non-run modes should be undone in `pm_power_state_exit_post_ops`.
The functions that do not need to be undone in `pm_power_state_exit_post_ops` should be put in `stm32_power_init`.